circleci: failed tests: TestReplicaRemovalCampaign

The following test appears to have failed:

[#19440](https://circleci.com/gh/cockroachdb/cockroach/19440):

```
I160622 13:08:51.134679 storage/replica_raftstorage.go:441  generated snapshot for range 1 at index 14 in 857.002µs. encoded size=3558, 31 KV pairs, 4 log entries
I160622 13:08:51.136348 storage/replica_raftstorage.go:533  replica 2 received snapshot for range 1 at index 14. encoded size=3558, 31 KV pairs, 4 log entries
I160622 13:08:51.137428 storage/replica_raftstorage.go:535  replica 2 applied snapshot for range 1 in 988.607µs
I160622 13:08:51.139488 storage/raft_transport.go:215  successfully sent MsgSnap to {2 2 2} at 127.0.0.1:33563
I160622 13:08:51.207575 storage/replica_command.go:2127  initiating a split of range=1 [/Min-/Max) at key "m"
panic: requested entry at index is unavailable

goroutine 8263 [running]:
panic(0x1cdca20, 0xc82017c990)
	/usr/local/go/src/runtime/panic.go:481 +0x3ff
github.com/coreos/etcd/raft.(*raftLog).term(0xc82088b490, 0xa, 0x6181f4, 0x0, 0x0)
	/go/src/github.com/coreos/etcd/raft/log.go:238 +0x1fe
github.com/coreos/etcd/raft.(*raftLog).lastTerm(0xc82088b490, 0xa)
	/go/src/github.com/coreos/etcd/raft/log.go:212 +0x57
github.com/coreos/etcd/raft.newRaft(0xc8216317f0, 0x0)
	/go/src/github.com/coreos/etcd/raft/raft.go:238 +0x114b
github.com/coreos/etcd/raft.NewRawNode(0xc8216317f0, 0x0, 0x0, 0x0, 0x7f6f203d00a8, 0x0, 0x0)
	/go/src/github.com/coreos/etcd/raft/rawnode.go:76 +0xe1
github.com/cockroachdb/cockroach/storage.(*Replica).withRaftGroupLocked(0xc8204cac40, 0xc821631998, 0x0, 0x0)
--
	/go/src/github.com/cockroachdb/cockroach/storage/replica_range_lease.go:118 +0x2c4
github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask.func1(0xc8205782a0, 0x29eadb6, 0x1e, 0x99, 0xc8208f1d60)
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:181 +0x74
created by github.com/cockroachdb/cockroach/util/stop.(*Stopper).RunAsyncTask
	/go/src/github.com/cockroachdb/cockroach/util/stop/stopper.go:182 +0xdc
FAIL	github.com/cockroachdb/cockroach/storage	17.549s
=== RUN   TestBatchBasics
I160622 13:09:06.512914 storage/engine/rocksdb.go:143  opening in memory rocksdb instance
I160622 13:09:06.515887 storage/engine/rocksdb.go:186  closing in-memory rocksdb instance
--- PASS: TestBatchBasics (0.02s)
=== RUN   TestBatchRepr
I160622 13:09:06.536341 storage/engine/rocksdb.go:143  opening in memory rocksdb instance
I160622 13:09:06.769199 storage/engine/rocksdb.go:186  closing in-memory rocksdb instance
--- PASS: TestBatchRepr (0.26s)
=== RUN   TestBatchGet
I160622 13:09:06.791121 storage/engine/rocksdb.go:143  opening in memory rocksdb instance

```
Please assign, take a look and update the issue accordingly.

--- a/server/node_test.go
+++ b/server/node_test.go
@@ -110,7 +110,7 @@ func createAndStartTestNode(addr net.Addr, engines []engine.Engine, gossipBS net
 	if err := node.start(addr, engines, roachpb.Attributes{}); err != nil {
 		t.Fatal(err)
 	}
-	if err := waitForInitialSplits(node.ctx.DB); err != nil {
+	if err := WaitForInitialSplits(node.ctx.DB); err != nil {
 		t.Fatal(err)
 	}
 	return grpcServer, addr, node, stopper

--- a/server/testserver.go
+++ b/server/testserver.go
@@ -276,10 +276,13 @@ func ExpectedInitialRangeCount() int {
 // splits at startup. If the expected range count is not reached within a
 // configured timeout, an error is returned.
 func (ts *TestServer) WaitForInitialSplits() error {
-	return waitForInitialSplits(ts.DB())
+	return WaitForInitialSplits(ts.DB())
 }
 
-func waitForInitialSplits(db *client.DB) error {
+// WaitForInitialSplits waits for the expected number of initial ranges to be
+// populated in the meta2 table. If the expected range count is not reached
+// within a configured timeout, an error is returned.
+func WaitForInitialSplits(db *client.DB) error {
 	expectedRanges := ExpectedInitialRangeCount()
 	return util.RetryForDuration(initialSplitsTimeout, func() error {
 		// Scan all keys in the Meta2Prefix; we only need a count.

--- a/storage/helpers_test.go
+++ b/storage/helpers_test.go
@@ -110,6 +110,18 @@ func (s *Store) LogReplicaChangeTest(txn *client.Txn, changeType roachpb.Replica
 	return s.logChange(txn, changeType, replica, desc)
 }
 
+// ReplicateQueuePurgatoryLength returns the number of replicas in replicate
+// queue purgatory.
+func (s *Store) ReplicateQueuePurgatoryLength() int {
+	return s.replicateQueue.PurgatoryLength()
+}
+
+// SetReplicaScannerDisabled turns replica scanning off or on as directed. Note
+// that while disabled, removals are still processed.
+func (s *Store) SetReplicaScannerDisabled(disabled bool) {
+	s.scanner.SetDisabled(disabled)
+}
+
 // GetLastIndex is the same function as LastIndex but it does not require
 // that the replica lock is held.
 func (r *Replica) GetLastIndex() (uint64, error) {

--- a/storage/queue.go
+++ b/storage/queue.go
@@ -222,6 +222,13 @@ func (bq *baseQueue) Length() int {
 	return bq.mu.priorityQ.Len()
 }
 
+// PurgatoryLength returns the current size of purgatory.
+func (bq *baseQueue) PurgatoryLength() int {
+	bq.mu.Lock()
+	defer bq.mu.Unlock()
+	return len(bq.mu.purgatory)
+}
+
 // SetDisabled turns queue processing off or on as directed.
 func (bq *baseQueue) SetDisabled(disabled bool) {
 	if disabled {

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -2307,12 +2307,16 @@ func (r *Replica) splitTrigger(
 		// Update store stats with difference in stats before and after split.
 		r.store.metrics.addMVCCStats(deltaMS)
 
-		// If the range was in not properly replicated before the split, the
-		// replicate queue may not have picked it up (due to the need for a
-		// split). Enqueue both new halves to speed up a potentially necessary
-		// replication. See #7022.
-		r.store.replicateQueue.MaybeAdd(r, r.store.Clock().Now())
-		r.store.replicateQueue.MaybeAdd(newRng, r.store.Clock().Now())
+		// If the range was not properly replicated before the split, the replicate
+		// queue may not have picked it up (due to the need for a split). Enqueue
+		// both new halves to speed up a potentially necessary replication. See
+		// #7022.
+		if len(split.UpdatedDesc.Replicas) == 1 {
+			r.store.replicateQueue.MaybeAdd(r, r.store.Clock().Now())
+		}
+		if len(split.NewDesc.Replicas) == 1 {
+			r.store.replicateQueue.MaybeAdd(newRng, r.store.Clock().Now())
+		}
 
 		// To avoid leaving the new range unavailable as it waits to elect
 		// its leader, one (and only one) of the nodes should start an

--- a/storage/replica_command.go
+++ b/storage/replica_command.go
@@ -2307,6 +2307,13 @@ func (r *Replica) splitTrigger(
 		// Update store stats with difference in stats before and after split.
 		r.store.metrics.addMVCCStats(deltaMS)
 
+		// If the range was in not properly replicated before the split, the
+		// replicate queue may not have picked it up (due to the need for a
+		// split). Enqueue both new halves to speed up a potentially necessary
+		// replication. See #7022.
+		r.store.replicateQueue.MaybeAdd(r, r.store.Clock().Now())
+		r.store.replicateQueue.MaybeAdd(newRng, r.store.Clock().Now())
+
 		// To avoid leaving the new range unavailable as it waits to elect
 		// its leader, one (and only one) of the nodes should start an
 		// election as soon as the split is processed.

--- a/storage/replicate_test.go
+++ b/storage/replicate_test.go
@@ -1,0 +1,47 @@
+// Copyright 2016 The Cockroach Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the License.
+//
+// Author: Peter Mattis (peter@cockroachlabs.com)
+
+package storage_test
+
+import (
+	"testing"
+
+	"github.com/cockroachdb/cockroach/server"
+	"github.com/cockroachdb/cockroach/util/leaktest"
+)
+
+func TestEagerReplication(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	store, stopper, _ := createTestStore(t)
+	defer stopper.Stop()
+
+	// Disable the replica scanner so that we rely on the eager replication code
+	// path that occurs after splits.
+	store.SetReplicaScannerDisabled(true)
+
+	if err := server.WaitForInitialSplits(store.DB()); err != nil {
+		t.Fatal(err)
+	}
+
+	// After the initial splits have been performed, all of the resulting ranges
+	// should be present in replicate queue purgatory (because we only have a
+	// single store in the test and thus replication cannot succeed).
+	expected := server.ExpectedInitialRangeCount()
+	if n := store.ReplicateQueuePurgatoryLength(); expected != n {
+		t.Fatalf("expected %d replicas in purgatory, but found %d",
+			expected, n)
+	}
+}

